tools/istio-iptables/pkg/capture/run.go

```diff
@@ -752,38 +752,156 @@ func (cfg *IptablesConfigurator) executeIptablesCommands(iptVer *dep.IptablesVer
+			log.Debugf("Expected iptables state: %#v", expectedState)
+			for table, chains := range expectedState {
+				_, ok := currentState[table]
+				if !ok {
+					deltaExists = true
+					log.Debugf("Can't find expected table %s in current state", table)
+					break check_loop
+				}
+				for chain, rules := range chains {
+					currentRules, ok := currentState[table][chain]
+					if !ok || (strings.HasPrefix(chain, "ISTIO_") && len(rules) != len(currentRules)) {
+						deltaExists = true
+						log.Debugf("Mismatching number of rules in chain %s between current and expected state", chain)
+						break check_loop
+					}
+				}
+			}
+			err = cfg.executeIptablesCommands(ipCfg.ver, ipCfg.checkRules)
+			if err != nil {
+				deltaExists = true
+				log.Debugf("iptables check rules failed")
+				break
+			}
 		}
+
+	}
+
+	if !residueExists {
+		log.Info("Clean-state detected, new iptables are needed")
+		return false, true
+	}
+
+	if deltaExists {
+		log.Warn("Found residues of old iptables rules/chains, reconciliation is needed")
 	} else {
-		// Execute iptables commands
-		if err := cfg.executeIptablesCommands(iptVer, cfg.ruleBuilder.BuildV4()); err != nil {
-			return err
+		log.Warn("Found compatible residues of old iptables rules/chains, reconciliation not needed")
+	}
+
+	return residueExists, deltaExists
+}
+
+func (cfg *IptablesConfigurator) executeCommands(iptVer, ipt6Ver *dep.IptablesVersion) error {
+	guardrails := false
+	defer func() {
+		if guardrails {
+			log.Info("Removing guardrails")
+			guardrailsCleanup := cfg.ruleBuilder.BuildCleanupGuardrails()
+			_ = cfg.executeIptablesCommands(iptVer, guardrailsCleanup)
+			_ = cfg.executeIptablesCommands(ipt6Ver, guardrailsCleanup)
 		}
-		// Execute ip6tables commands
-		if err := cfg.executeIptablesCommands(ipt6Ver, cfg.ruleBuilder.BuildV6()); err != nil {
-			return err
+	}()
+
+	residueExists, deltaExists := cfg.VerifyIptablesState(iptVer, ipt6Ver)
+	if residueExists && deltaExists && !cfg.cfg.Reconcile {
+		return fmt.Errorf("reconcile is needed but no-reconcile flag is set. Can't recover from this state")
+	}
+	// Cleanup Step
+	if (residueExists && deltaExists) || cfg.cfg.CleanupOnly {
+		// Apply safety guardrails
+		if !cfg.cfg.CleanupOnly {
+			log.Info("Setting up guardrails")
+			guardrailsCleanup := cfg.ruleBuilder.BuildCleanupGuardrails()
+			guardrailsRules := cfg.ruleBuilder.BuildGuardrails()
+			for _, ver := range []*dep.IptablesVersion{iptVer, ipt6Ver} {
+				cfg.tryExecuteIptablesCommands(ver, guardrailsCleanup)
+				if err := cfg.executeIptablesCommands(ver, guardrailsRules); err != nil {
+					return err
+				}
+				guardrails = true
+			}
+		}
+		// Remove old iptables
+		log.Info("Performing cleanup of existing iptables")
+		cfg.tryExecuteIptablesCommands(iptVer, cfg.ruleBuilder.BuildCleanupV4())
+		cfg.tryExecuteIptablesCommands(ipt6Ver, cfg.ruleBuilder.BuildCleanupV6())
+	}
+
+	// Apply Step
+	if deltaExists && !cfg.cfg.CleanupOnly {
```

Comment:
Can we keep the backwards compatible behavior of always running it?

Reply:
Done

Reply:
This is still not done, its skipping the run when deltaExists == false, which may be a false positive?

Reply:
Sorry, I thought the comment was related to "fail if no reconciliation and pre-existing iptables are found" and marked it as solved.

The"skip if delta exists" is intended to prevent reruns when all iptables are detected as already in place, such as in istio-init reruns. I cannot force a rerun exactly as it was before this commit because that behavior is problematic.
Perhaps, if we detect no delta, we could still rerun the apply step but ignore the error that will likely be raised. While not very clean, this approach would be more backward compatible. WTDY?

On the other end, a false negative on deltaExists would be difficult because it would require meeting two conditions:
- Having the same number of rules in every chain between the "expected state" and "current state"
- Having no failures in any of the "iptables -C ..." commands that are executed
If there's an issue in the parser, it's more likely to lead to missfire on reconciliation (residueExists && deltaExists) rather than this skip.

Reply:
We can keep it on by default but at least should have a feature flag to turn it off, since this is very tricky behavior. We have ~never made an iptables change that didn't have a bug in it in some obscure environment, so probably there is some weird edge cases that this will matter. Having a way to opt-out of that would be a good safety mechanism